File: ruby/card-raytracer.rb
WIDTH  = 512
HEIGHT = 512

class Vector
  def initialize(x, y, z)
    @x = x
    @y = y
    @z = z
  end

  def x
    @x
  end

  def y
    @y
  end

  def z
    @z
  end

  def +(a)
    return Vector.new(@x + a.x, @y + a.y, @z + a.z)
  end

  def *(a)
    return Vector.new(@x * a, @y * a, @z * a)
  end

  def %(a)
    return @x * a.x + @y * a.y + @z * a.z
  end

  def ^(a)
    return Vector.new(@y * a.z - @z * a.y, @z * a.x - @x * a.z,
      @x * a.y - @y * a.x)
  end

  def !()
    return self * (1.0 / Math.sqrt(self % self))
  end
end

# Position vectors:
ZERO_VECTOR = Vector.new(0.0, 0.0, 0.0)
Z_ORTHO_VECTOR = Vector.new(0.0, 0.0, 1.0)
CAMERA_ASPECT_VECTOR = Vector.new(17.0, 16.0, 8.0)
CAMERA_DEST_VECTOR = Vector.new(-6.0, -16.0, 0.0)
# Color vectors:
COLOR_CELL1_VECTOR = Vector.new(3.0, 1.0, 1.0)
COLOR_CELL2_VECTOR = Vector.new(3.0, 3.0, 3.0)
COLOR_DARK_GRAY_VECTOR = Vector.new(13.0, 13.0, 13.0)
COLOR_SKY = Vector.new(0.7, 0.6, 1.0)

G = [ 
  0x0003C712,  # 00111100011100010010
  0x00044814,  # 01000100100000010100
  0x00044818,  # 01000100100000011000
  0x0003CF94,  # 00111100111110010100
  0x00004892,  # 00000100100010010010
  0x00004891,  # 00000100100010010001
  0x00038710,  # 00111000011100010000
  0x00000010,  # 00000000000000010000
  0x00000010   # 00000000000000010000
]

class TracerOutput
  def initialize(m, t, n)
    @m = m
    @t = t
    @n = n
  end

  def m
    @m
  end

  def t
    @t
  end

  def n
    @n
  end

  def m=(m)
    @m = m
  end

  def t=(t)
    @t = t
  end

  def n=(n)
    @n = n
  end
end

def tracer(o, d, out)
  t = 1.0e9
  m = 0
  n = out.n
  p = -o.z / d.z
  if 0.01 < p
    t = p
    n = Z_ORTHO_VECTOR
    m = 1
  end
  18.downto 0 do |k|
    8.downto 0 do |j|
      if G[j] & 1 << k != 0
        p = o + Vector.new(-k.to_f, 0.0, -j.to_f - 4.0)
        b = p % d
        c = p % p - 1
        q = b * b - c
        if q > 0.0
          s = -b - Math.sqrt(q)
          if s < t && s > 0.01
            t = s
            n = !(p + d * t)
            m = 2
          end
        end
      end
    end
  end
  out.m = m
  out.t = t
  out.n = n
end

def sampler(o, d)
  t = 0.0
  n = ZERO_VECTOR
  m = 0
  out = TracerOutput.new(m, t, n)
  tracer(o, d, out)
  m = out.m
  t = out.t
  n = out.n
  if m == 0
    return COLOR_SKY * ((1.0 - d.z) ** 4)
  end
  h = o + d * t
  l = !(Vector.new(9.0 + rand, 9.0 + rand, 16.0) + h * -1.0)
  r = n

  b = l % n
  if b < 0.0
    b = 0.0
  else
    m1 = 0
    out = TracerOutput.new(m1, t, n)
    tracer(h, l, out)
    m1 = out.m
    t = out.t
    n = out.n
    if m1 != 0
      b = 0.0
    end
  end

  if m == 1
    return (((h.x * 0.2).ceil + (h.y * 0.2).ceil).to_i & 1 != 0 ? 
      COLOR_CELL1_VECTOR : COLOR_CELL2_VECTOR) * (b * 0.2 + 0.1)
  end

  r = d + r * (r % d * -2.0)
  p = (b > 0.0) ? ((l % r) ** 99) : 0.0
  return Vector.new(p, p, p) + sampler(h, r) * 0.5
end


if ARGV.empty?
  STDERR.puts "\n\nUsage: card-raytracer <filename>.ppm\n"
  exit -1
end
File.open(ARGV[0], "w") do |out|
  out.printf("P6 %d %d 255 ", WIDTH, HEIGHT)
  g = !CAMERA_DEST_VECTOR
  a = !(Z_ORTHO_VECTOR ^ g) * 0.002
  b = !(g ^ a) * 0.002
  c = (a + b) * -256.0 + g
  (HEIGHT-1).downto 0 do |y|
    (WIDTH-1).downto 0 do |x|
      p = COLOR_DARK_GRAY_VECTOR
      63.downto 0 do |r|
        t = a * (rand - 0.5) * 99.0 + b * (rand - 0.5) * 99.0
        p = p + sampler(
          CAMERA_ASPECT_VECTOR + t,
          !(t * -1.0 + (a * (rand + x) + b * (y + rand) + c) * 16.0)
          ) * 3.5
      end
      out.print p.x.to_i.chr, p.y.to_i.chr, p.z.to_i.chr
    end
  end
end

File: java/CardRaytracer.java
import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.Duration;
import java.time.Instant;

import static java.lang.Integer.valueOf;
import static java.lang.Math.*;
import static java.lang.System.err;

/**
 * Special thanks to Paul Heckbert
 * <p/>
 * http://tproger.ru/translations/business-card-raytracer
 * <p/>
 * Java port
 * <p/>
 * 4-SEP-2015 - (mayton) In beginning...
 * 5-SEP-2015 - (mayton) changes
 * 6-SEP-2015 - (mayton) changes
 * 2-MAR-2016 - (mayton) constants up to 1.7 JDK
 * 20-JUN-2016 - (mayton) I/O has been improoved acc. to Java8.
 *                        System clock usage has been inprooved acc. to Java8
 */

public class CardRaytracer {

    class VectorBox {

        public Vector value;

        public VectorBox(Vector value) {
            this.value = value;
        }
    }

    class DoubleBox {

        public double value;

        public DoubleBox(double value) {
            this.value = value;
        }
    }

    static final class Vector {

        public final double x, y, z;

        public Vector(double x, double y, double z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public Vector sum(Vector r) {
            return new Vector(x + r.x, y + r.y, z + r.z);
        }

        public Vector prod(double r) {
            return new Vector(x * r, y * r, z * r);
        }

        public double sprod(Vector r) {
            return x * r.x + y * r.y + z * r.z;
        }

        public Vector vprod(Vector r) {
            return new Vector(y * r.z - z * r.y,
                    z * r.x - x * r.z,
                    x * r.y - y * r.x);
        }

        public Vector norm() {
            // return *this * (1 / sqrt(*this % *this));
            return this.prod(1.0 / sqrt(this.sprod(this)));
        }

    }


    static final int WIDTH = 512;
    static final int HEIGHT = 512;
    static final int SUB_SAMPLES = 64;

    // Position vectors:
    static final Vector ZERO_VECTOR            = new Vector(0.0,0.0,0.0);
    static final Vector Z_ORTHO_VECTOR         = new Vector(0.0,0.0,1.0);
    static final Vector CAMERA_ASPECT_VECTOR   = new Vector(17.0, 16.0, 8.0);
    static final Vector CAMERA_DEST_VECTOR     = new Vector(-6.0, -16.0, 0.0);

    // Color vectors:
    static final Vector COLOR_CELL1_VECTOR     = new Vector(3.0, 1.0, 1.0);
    static final Vector COLOR_CELL2_VECTOR     = new Vector(3.0, 3.0, 3.0);
    static final Vector COLOR_DARK_GRAY_VECTOR = new Vector(13.0, 13.0, 13.0);
    static final Vector COLOR_SKY              = new Vector(0.7, 0.6, 1.0);

    int width  = WIDTH;
    int height = HEIGHT;

    OutputStream out;

    static final int G[] = {
            0b00111100011100010010,
            0b01000100100000010100,
            0b01000100100000011000,
            0b00111100111110010100,
            0b00000100100010010010,
            0b00000100100010010001,
            0b00111000011100010000,
            0b00000000000000010000,
            0b00000000000000010000
    };

    static final int M = 1_048_576; // 2^20
    static final int J = 2_045;
    int oldI = 12357;

    double Random() {
        oldI = (oldI * J + 1) % M;
        return (double) oldI / M;
    }

    int tracer(final Vector o,final Vector d,final DoubleBox t,final VectorBox n) {
        t.value = 1e9;
        int m = 0;
        double p = -o.z / d.z;

        if (0.01 < p) {
            t.value = p;
            n.value = Z_ORTHO_VECTOR;
            m = 1;
        }

        for (int k = 18; k >= 0; k--) {
            for (int j = 8; j >= 0; j--) {
                // if (G[j] & 1 << k)
                if ((G[j] & 1 << k) != 0) {
                    // Vector p = o + Vector(-k, 0, -j - 4);
                    Vector _p = o.sum(new Vector(-k, 0.0, -j - 4.0));
                    // double b = p % d;
                    double b = _p.sprod(d);
                    // double c = p % p - 1;
                    double c = _p.sprod(_p) - 1.0;
                    // double q = b * b - c;
                    double q = b * b - c;
                    if (q > 0) {
                        double s = -b - sqrt(q);
                        if (s < t.value && s > 0.01) {
                            t.value = s;
                            // n = !(p + d * t);
                            n.value = (_p.sum(d.prod(t.value))).norm();
                            m = 2;
                        }
                    }
                }
            }
        }
        return m;
    }


    /**
     *
     * @param o
     * @param d
     * @return Color of sample
     */
    Vector sampler(final Vector o,final Vector d) {
        DoubleBox t = new DoubleBox(0.0);
        VectorBox n = new VectorBox(ZERO_VECTOR);
        int m = tracer(o, d, t, n);

        if (m == 0) {
            // return Vector(.7, .6, 1) * pow(1 - d.z, 4);
            return COLOR_SKY.prod(pow(1.0 - d.z, 4));
        }
        // Vector h = o + d * t;
        Vector h = o.sum(d.prod(t.value));

        // Vector l = !(Vector(9 + Random(), 9 + Random(), 16) + h * -1);
        Vector l = new Vector(9.0 + Random(), 9.0 + Random(), 16.0).sum(h.prod(-1.0)).norm();

        // Vector r = d + n * (n % d * -2);
        Vector r = d.sum(n.value.prod(n.value.sprod(d) * -2.0));
        // double b = l % n;
        double b = l.sprod(n.value);
        if (b < 0 || tracer(h, l, t, n) != 0) {
            b = 0;
        }
        // double p = pow(l % r * (b > 0), 99);
        double p = pow(l.sprod(r) * (b > 0 ? 1.0 : 0.0), 99.0);
        if ((m & 1) != 0) {
            // h = h * .2;
            h = h.prod(0.2);
            // return ((int) (ceil(h.x) + ceil(h.y)) & 1 ?
            //              Vector(3, 1, 1) :
            //              Vector(3, 3, 3)) * (b * .2 + .1);
            return ((((int) (ceil(h.x) + ceil(h.y)) & 1) != 0) ?
                    COLOR_CELL1_VECTOR :
                    COLOR_CELL2_VECTOR).prod(b * 0.2 + 0.1);
        }
        // return Vector(p, p, p) + sampler(h, r) * .5;
        return new Vector(p, p, p).sum(sampler(h, r).prod(0.5));
    }

    public CardRaytracer(OutputStream out, int width, int height) {
        this.width = width;
        this.height = height;
        this.out = out;
    }

    public CardRaytracer(OutputStream out) {
        this.out = out;
    }

    public void process() throws IOException {
        PrintWriter wr = new PrintWriter(out);
        wr.printf("P6 %d %d 255 ", width, height);
        wr.flush();
        // Vector g = !Vector(-6, -16, 0);
        Vector g = CAMERA_DEST_VECTOR.norm();
        // Vector a = !(Vector(0, 0, 1) ^ g) * .002;
        Vector a = (Z_ORTHO_VECTOR.vprod(g)).norm().prod(0.002);
        // Vector b = !(g ^ a) * .002;
        Vector b = g.vprod(a).norm().prod(0.002);
        // Vector c = (a + b) * -256 + g;
        Vector c = a.sum(b).prod(-256.0).sum(g);
        for (int y = height - 1; y >= 0; y--) {
            for (int x = width - 1; x >= 0; x--) {
                Vector p = COLOR_DARK_GRAY_VECTOR;
                for (int r = 0; r < SUB_SAMPLES; r++) {
                    //Vector t = a * (Random() - .5) * 99 + b * (Random() - .5) * 99;
                    Vector t = a.prod(Random() - 0.5).prod(99.0).sum(b.prod(Random() - 0.5).prod(99.0));
                    //p = sampler(Vector(17, 16, 8) + t,
                    //           !(t * -1 + (a * (Random() + x) + b * (y + Random()) + c) * 16)) * 3.5 + p;
                    p = sampler(CAMERA_ASPECT_VECTOR.sum(t),
                            t.prod(-1.0).sum(a.prod(Random() + x).sum(b.prod(Random() + y)).sum(c).prod(16.0)).norm()
                    ).prod(3.5).sum(p);
                }
                int R = (int) p.x;
                int G = (int) p.y;
                int B = (int) p.z;
                out.write(R);
                out.write(G);
                out.write(B);
            }
        }
        out.flush();
    }

    public static void main(String[] args) throws Exception {
        Instant start = Instant.now();
        if (args.length == 0) {
            new CardRaytracer(System.out).process();
            err.printf("Elapsed time : %d sec\n",Duration.between(start,Instant.now()).toMillis() / 1000);
            System.exit(1);
        } else if (args.length == 1) {
            OutputStream os = Files.newOutputStream(Paths.get(args[0]));
            new CardRaytracer(os).process();
            err.printf("Elapsed time : %d sec\n",Duration.between(start,Instant.now()).toMillis() / 1000);
            System.exit(2);
        } else if (args.length == 3) {
            OutputStream os = Files.newOutputStream(Paths.get(args[0]));
            new CardRaytracer(os, valueOf(args[1]), valueOf(args[2])).process();
            err.printf("Elapsed time : %d sec\n",Duration.between(start,Instant.now()).toMillis() / 1000);
            System.exit(2);
        }
        System.exit(0);
    }

}
